internal/api/core/middleware/proxy.go

```diff
@@ -1,118 +1,226 @@
+	if !globalDatasourceMatch && !localDatasourceMatch {
+		// this is likely a request for the API itself
+		return nil, "", nil
+	}
+
+	if globalDatasourceMatch {
+		return getGlobalDatasourceAndPath(globalDTS, requestPath)
+	}
+	return getLocalDatasourceAndPath(dts, requestPath)
+}
+
+func getGlobalDatasourceAndPath(dao globaldatasource.DAO, requestPath string) (v1.DatasourceSpec, string, error) {
+	matchingGroups := globalProxyMatcher.FindAllStringSubmatch(requestPath, -1)
 	if len(matchingGroups) > 1 || len(matchingGroups[0]) <= 1 {
-		return nil, echo.NewHTTPError(http.StatusBadGateway, "unable to forward the request to the datasource, request not properly formatted")
+		return nil, "", echo.NewHTTPError(http.StatusBadGateway, "unable to forward the request to the datasource, request not properly formatted")
 	}
 	datasourceName := matchingGroups[0][1]
 	// getting the datasource object
 	dts, err := dao.Get(datasourceName)
 	if err != nil {
 		if etcd.IsKeyNotFound(err) {
 			logrus.Debugf("unable to find the Datasource '%s'", datasourceName)
-			return nil, echo.NewHTTPError(http.StatusBadGateway, fmt.Sprintf("unable to forward the request to the datasource '%s', datasource doesn't exist", datasourceName))
+			return nil, "", echo.NewHTTPError(http.StatusBadGateway, fmt.Sprintf("unable to forward the request to the datasource '%s', datasource doesn't exist", datasourceName))
 		}
 		logrus.WithError(err).Errorf("unable to find the datasource '%s', something wrong with the database", datasourceName)
-		return nil, echo.NewHTTPError(http.StatusInternalServerError, "internal server error")
+		return nil, "", echo.NewHTTPError(http.StatusInternalServerError, "internal server error")
 	}
 	// Based on the HTTP 1.1 RFC, a `/` should be the minimum path.
 	// https://datatracker.ietf.org/doc/html/rfc2616#section-5.1.2
 	path := "/"
 	if len(matchingGroups[0]) > 1 {
 		path = matchingGroups[0][2]
 	}
+	return dts.Spec, path, nil
+}
 
-	// redirect the request to the datasource
-	req.URL.Path = path
-	return dts.Spec.URL, nil
+func getLocalDatasourceAndPath(dao datasource.DAO, requestPath string) (v1.DatasourceSpec, string, error) {
+	matchingGroups := globalProxyMatcher.FindAllStringSubmatch(requestPath, -1)
+	if len(matchingGroups) > 1 || len(matchingGroups[0]) <= 2 {
+		return nil, "", echo.NewHTTPError(http.StatusBadGateway, "unable to forward the request to the datasource, request not properly formatted")
+	}
+	projectName := matchingGroups[0][1]
+	datasourceName := matchingGroups[0][2]
+	// getting the datasource object
+	dts, err := dao.Get(projectName, datasourceName)
+	if err != nil {
+		if etcd.IsKeyNotFound(err) {
+			logrus.Debugf("unable to find the Datasource '%s' in project '%s'", datasourceName, projectName)
+			return nil, "", echo.NewHTTPError(http.StatusBadGateway, fmt.Sprintf("unable to forward the request to the datasource '%s', datasource doesn't exist", datasourceName))
```

Comment:
nit: Should this be http.StatusNotFound?

Reply:
that's a good question. Looking at the definition of the error `BadGateway` probably it's indeed better to return the status not found